[Inductor] fix alignement assumption for fallback

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150804
* __->__ #150777

Inductor right now only works properly for fallback kernels producing aligned output.
When Inductor create layout for fallback kernel output, Inductor does not add the tensor offset to the layout [link](https://github.com/pytorch/pytorch/blob/2a1e2b88ed7bf7d7436b741ee0c3a2297d7d7bc2/torch/_inductor/ir.py#L6935-L6941). Thus unaligned output will be treated as aligned. Adding the offset to the layout directly does not work since that change the index expression in the generated kernel and we may 'double' applying the offset. Triton already considers the offset when passing in the data_ptr.

To solve this issue, we track the unaligned buffer names instead.

This potentially can fix the internal issues we are debugging here: https://fb.workplace.com/groups/1075192433118967/permalink/1618308128807392/

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D72600784](https://our.internmc.facebook.com/intern/diff/D72600784)

shunting314 committed:
[Inductor] fix alignement assumption for fallback

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -1397,6 +1397,8 @@ def output(self, name: str) -> str:
         return self._lookup("out_ptr", self.output_buffers, name)
 
     def make_inplace(self, input_name: str, output_name: str) -> None:
+        if input_name in V.graph.unaligned_buffers:
+            V.graph.unaligned_buffers.add(output_name)
```
What about buffer reuse ? I suppose with graph inputs we have not had to worry about this as much because we dont reuse them quite as often.

shunting314 replied:
Easiest way is to add alignment predicate into the buffer reuse key. Other thoughts?

eellison replied:
or just not re-use the buffer if it's unaligned, since access should be slower and it is a marginal gain there anyway.

eellison reviewed (commented):
Looks good, although I wonder how consistently the metas will be written correctly.. also wonder if we need more robust buffer reuse / aliasing mechanism here..

shunting314 commented:
> although I wonder how consistently the metas will be written correctly..

The PR assumes the meta is correct. But it's a good call. I'll add an env-var so we can conservatively assume fallback output is non-aligned. The env-var can be used for testing

shunting314 commented:
btw, I didn't add assertion for alignment yet since I see we can repro the problem even with correct meta. Not sure if this is something that happens in practice

shunting314 commented on torch/_inductor/utils.py:
```diff
@@ -2465,6 +2465,15 @@ def copy_misaligned_inputs(
             new_inputs[i] = clone_preserve_strides(_inp)
 
 
+def is_tensor_aligned(tensor: torch.Tensor) -> bool:
```
I can reuse this one

shunting314 committed:
Update on "[Inductor] fix alignement assumption for fallback"


Inductor right now only works properly for fallback kernels producing aligned output.
When Inductor create layout for fallback kernel output, Inductor does not add the tensor offset to the layout [link](https://github.com/pytorch/pytorch/blob/2a1e2b88ed7bf7d7436b741ee0c3a2297d7d7bc2/torch/_inductor/ir.py#L6935-L6941). Thus unaligned output will be treated as aligned. Adding the offset to the layout directly does not work since that change the index expression in the generated kernel and we may 'double' applying the offset. Triton already considers the offset when passing in the data_ptr.

To solve this issue, we track the unaligned buffer names instead.

This potentially can fix the internal issues we are debugging here: https://fb.workplace.com/groups/1075192433118967/permalink/1618308128807392/

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72600784).

shunting314 commented:
@pytorchbot merge
